### PR TITLE
KAFKA-14455: Kafka Connect create and update REST APIs should surface failures while writing to the config topic

### PR DESCRIPTION
**Pull request by yashmayya**
- Kafka Connect's `POST /connectors` and `PUT /connectors/{connector}/config` REST APIs internally simply write a message to the Connect cluster's internal config topic (which is then processed asynchronously by the herder). 
- However, no callback is passed to the producer's send method and there is no error handling in place for producer send failures (see [here](https://github.com/apache/kafka/blob/c1a54671e8fc6c7daec5f5ec3d8c934be96b4989/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java#L716) / [here](https://github.com/apache/kafka/blob/c1a54671e8fc6c7daec5f5ec3d8c934be96b4989/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java#L726)).
- Consider one such case where the Connect worker's principal doesn't have a `WRITE` ACL on the cluster's config topic. Now suppose the user submits a connector's configs via one of the above two APIs. The producer send [here](https://github.com/apache/kafka/blob/c1a54671e8fc6c7daec5f5ec3d8c934be96b4989/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java#L716) / [here](https://github.com/apache/kafka/blob/c1a54671e8fc6c7daec5f5ec3d8c934be96b4989/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java#L726) won't succeed (due to a `TopicAuthorizationException`) but the API responses will be `201 Created` success responses anyway.
- This is a very poor UX because the connector will actually never be created but the API response would indicate a successful operation. Furthermore, this failure would only be detectable if TRACE logs are enabled (via [this log)](https://github.com/apache/kafka/blob/df29b17fc40f7c15460988d58bc652c3d66b60f8/clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java) making it near impossible for users to debug.
- This PR updates the `KafkaConfigBackingStore` APIs to use synchronous producer sends in the non-EOS enabled case which surfaces exceptions and also brings the behavior in-line with the EOS enabled case (where the transaction commit ensures the same behavior).
- Side quest: update the outdated Javadoc for `KafkaConfigBackingStore` and improve readability of its rendered form.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java
Comment (yashmayya) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: Another option could be to directly call `callback.onCompletion` in `KafkaConfigBackingStore::putConnectorConfig` itself - i.e. construct the `Created<ConnectorInfo>` object there using its in-memory maps directly rather than the snapshot; this would avoid this double callback kinda mechanism but would need some other refactoring.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConfigBackingStore.java
Comment (yashmayya) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConfigBackingStore.java]: Could potentially add a new overloaded method in `KafkaConfigBackingStore` and avoid touching this interface method although that would require refactoring `AbstractHerder` to make it generic in order to allow `DistributedHerder` to access an instance of `KafkaConfigBackingStore` rather than `ConfigBackingStore`. However, that would be a much more noisy change IMO.

**Comment by yashmayya** [comment]
@C0urante could you please take a look at this whenever you get a chance? I'm sorry to keep pinging you for review requests, but I'm not aware of any other committers currently taking a look at Connect PRs :)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: this comment is stale and can be removed
(i forgot to remove it sorry)
```suggestion
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: The only other call-site for the send(String, byte[]) is now putTargetState, does a similar fix apply there?
Perhaps this blind send call is not safe and should be eliminated. Perhaps we can also enforce non-null callbacks. WDYT?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: I believe here we're intentionally relying on this behavior of commitTransaction to propagate errors:
> Further, if any of the {@link #send(ProducerRecord)} calls which were part of the transaction hit irrecoverable
> errors, this method will throw the last received exception immediately and the transaction will not be committed.
> So all {@link #send(ProducerRecord)} calls in a transaction must succeed in order for this method to succeed.

I'm not sure if we're in danger of double-completing the callback, and whether semantically the callback should be completed after the commitTransaction completes or an exception is thrown from commitTransaction.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: Thanks, that's a good observation. Currently, the pause / resume APIs (which call `putTargetState`) don't use callback mechanisms whatsoever and they also don't have any documented response bodies (they send back a `202` response with an empty body right now) and will hence need a bit more refactoring. I feel like that should be done in a follow up / separate PR, WDYT?

Edit: Actually, on second thought it's a closely related issue and could be taken care of in this PR itself. I'll work on making the requisite changes.

Edit 2: Although now with these changes, there's an inconsistency between different APIs since `deleteConnectorConfig`, `putTaskConfigs` and a couple of other methods are still using `null` callbacks.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: > I believe here we're intentionally relying on this behavior of commitTransaction to propagate errors:


Thanks for pointing this out!

As per https://kafka.apache.org/33/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#send-

```
When used as part of a transaction, it is not necessary to define a callback or check the result of the future in order to detect errors from send. If any of the send calls failed with an irrecoverable error, the final [commitTransaction()](https://kafka.apache.org/33/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#commitTransaction--) call will fail and throw the exception from the last failed send. When this happens, your application should call [abortTransaction()](https://kafka.apache.org/33/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#abortTransaction--) to reset the state and continue to send data.
```

So what you're saying definitely does make sense, and I tried this scenario out manually - for an EOS enabled worker, the response from the `POST /connectors` API when the worker's principal doesn't have a `WRITE` ACL on the config topic is (without the changes from this PR): 

```
{
  "error_code": 500,
  "message": "Failed to write to config topic; this may be due to a transient error and the request can be safely retried"
}
```

While the worker logs do have the exact root cause as well (the `TopicAuthorizationException`), I believe the REST API response in this case isn't all that helpful to the user. With the changes from this PR, the response in the same scenario looks like:

```
{
  "error_code": 500,
  "message": "Not authorized to access topics: [connect-configs]"
}
```

> I'm not sure if we're in danger of double-completing the callback

I don't believe we are in danger of double completing the callback, although you're right in that we are unnecessarily handling the error in two places in a way.

We could avoid the use of callbacks in the producer send (although I couldn't find anything that explicitly warns against doing so) when using the transactional producer and instead refactor [this](https://github.com/apache/kafka/blob/e3585a4cd5ddb5b8475a49c38143d18e7c640bfe/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L1494) / [this](https://github.com/apache/kafka/blob/e3585a4cd5ddb5b8475a49c38143d18e7c640bfe/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java#L731) so that the final exception that is thrown is the producer exception itself rather than the (doubly) wrapped one. WDYT?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: > Edit 2: Although now with these changes, there's an inconsistency between different APIs since deleteConnectorConfig, putTaskConfigs and a couple of other methods are still using null callbacks.

Rather than rewriting all of the `ConfigBackingStore` methods to accept callbacks (making them look async), maybe we can terminate the producer callbacks in the `KafkaConfigBackingStore` and keep the `ConfiBackingStore` methods synchronous. I think this has other advantages, especially for the `putConnectorConfig`, where we can enforce that the producer send completes before the `readToEnd()`.

It appears that the original signatures for the `ConfigBackingStore` provided a synchronous API. The javadocs don't mention what exceptions are thrown, but there are some methods which already throw exceptions, such as those coming from the readToEnd() calls. We could certainly throw exceptions rather than propagate them via callbacks.

In fact, the only `*BackingStore` method which was already async was `OffsetBackingStore::set(Map, Callback)`, which appears was only written that way to log errors (no change in control flow, retries, etc). That method is now involved in some callback hell in ConnectorOffsetBackingStore used for EOS source that could also be rewritten to be synchronous. (I don't think this is in-scope for this PR and is just an interesting lateral refactor)

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Yes, we should certainly try to propagate the more informative error message. Maybe we can find a control flow where the callback error is propagated before calling the commitTransaction and getting the more generic error message.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: I agree that terminating the producer callbacks in the `KafkaConfigBackingStore` itself would look cleaner. However, if we want to surface producer errors to users via the REST API responses (which is the goal for this PR), we would need to complete the request callbacks either directly or via an intermediary callback (also see [this comment](https://github.com/apache/kafka/pull/12984#discussion_r1046720576)) in the `KafkaConfigBackingStore`  (both of which would involve changing the method signatures to accept callbacks) due to the way that the REST API responses are generated using `FutureCallback` in `ConnectorsResource`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: I believe that `commitTransaction` ensures that all producer records are flushed and all pending callbacks invoked before the transaction is committed, so this - `where the callback error is propagated before calling the commitTransaction and getting the more generic error message.` should already be the case with the current changes?

However, it doesn't look like using callbacks with the transactional producer offers much benefit - we could simply reword the existing exception message (`this may be due to a transient error and the request can be safely retried`) to indicate that it could potentially denote a non-transient error as well. Furthermore, the specific case that this PR attempted to fix (missing WRITE ACL on the config topic not being surfaced to users properly) is anyway highly unlikely to go unnoticed in an EOS enabled Connect cluster since the herder thread itself will repeatedly hit this condition [here](https://github.com/apache/kafka/blob/9d1f9f77642d7e95dec37647657478ed187bdeb7/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L419) in the tick loop (the worker logs will reveal the underlying `TopicAuthorizationException`) and request processing won't happen at all (all external requests that are run on the herder's thread will timeout).

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: > if we want to surface producer errors to users ... which would involve changing the method signatures to accept callbacks

This is not true; you can surface errors synchronously by throwing exceptions inside the `KafkaConfigBackingStore`  and catching them in the caller. This is what I meant by terminating the producer callbacks: transform the control flow from asynchronous via a callback to synchronous results.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: > should already be the case with the current changes

It may be the case but not for the reason you specified. The guarantee from the producer is that the record callbacks are completed before the commitTransaction _returns_, while I was suggesting a control flow where we ensure that record callbacks are completed before the commitTransaction is _called_. If you still call commitTransaction, you will get both the callback's error and the synchronous error from commitTransaction, and would still need to make sure that the correct exception is shadowed. The control over this shadowing is handled by this code and whoever is terminating the callback in the REST layer.

> we could simply reword the existing exception message

This is certainly the easy way out, but I don't think we should settle for that. Vague error messages indicate that they aren't capturing and reporting the exception properly, and making them more vague to compensate doesn't help anyone diagnose the underlying issue. If we can make a control flow change that reports the errors faithfully, that is going to help someone down the line debugging it.

> the worker logs will reveal the underlying TopicAuthorizationException

Is this the case? The worker is using _this_ code path to write the session key; If we're hiding the result from the REST calls, are we not also hiding the error from the herder tick thread?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: Oh okay, I see what you're saying now. Sorry, I think I misunderstood you earlier. I agree with you, none of the `KafkaConfigBackingStore` use cases seem to necessarily require the performance benefit of using async send with callbacks. Although to make it synchronous, I think it might be better to avoid using the producer callback altogether and instead call `get()` on the returned future (which throws any exceptions that were encountered while sending the record), WDYT?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: I'm not sure I follow what benefit we'd be getting here by handling both the producer callback error as well as the one thrown by `commitTransaction`? The control flow would be more straightforward by removing the producer callback and just relying on `commitTransaction` to throw exceptions, if any. The producer's Javadoc itself also suggests that callbacks need not be defined when using the transactional producer since `commitTransaction` will throw the error from the last failed send in a transaction.

> making them more vague to compensate

We wouldn't be making it more vague. The message would state that the write to the config topic failed which is the cause for failure. Since the exception mapper used by Connect's REST server only writes the [top level exception's message](https://github.com/apache/kafka/blob/d798ec779c25dba31fa5ee9384d159ed54c6e07b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/errors/ConnectExceptionMapper.java#L72) to the response (i.e. nested exceptions aren't surfaced via the REST API response), I think it makes sense to keep the top level exception's message generic and allow users to debug further via the worker logs (where the entire exception chain's stack trace will be visible). Note that I'm suggesting a similar change for the non-EOS enabled case as well - i.e. don't use the producer error directly [here](https://github.com/apache/kafka/pull/12984/files#diff-c346b7b90fe30ac08b4211375fe36208139a90e40838dd3a7996021a8c4c5b13R1064), instead wrapping it in a `ConnectException` which says that the write to the config topic failed. The reasoning here is that since a Connect user may not even know that Connect uses a producer under the hood to write certain requests to the config topic for asynchronous processing, it would make more sense to have an informative Connect specific exception message rather than directly throwing the producer exception which may or may not contain enough details to be relevant to a Connect user.

>  If we're hiding the result from the REST calls, are we not also hiding the error from the herder tick thread?

Hm no, the hiding issue was only for non-EOS enabled workers. Like I've pointed out above, for workers that have EOS enabled, the REST API does return a `500` response.

Edit: Another option for the above issue could be changing the exception mapper to concatenate all the exception messages from the exception chain.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: I've gone ahead and made the changes to convert the `KafkaConfigBackingStore` APIs to be synchronous even when EOS is disabled (thus making the behavior more in-line with the EOS case). This simplifies things significantly while also providing consistency w.r.t error handling across all the `KafkaConfigBackingStore` APIs without making invasive changes.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: I've removed the usage of the producer callback here since we're moving to synchronous usage of producer send in the non-EOS case as well anyway (aside from the earlier point that it doesn't really make sense to handle errors via both a callback as well as `commitTransaction`). The behavior of surfacing exceptions synchronously is similar in both cases now; one through calling `get()` on the returned future from `Producer::send` and the other through `Producer::commitTransaction`.

> Another option for the above issue could be changing the exception mapper to concatenate all the exception messages from the exception chain.

Yet another option for this could be to simply append a "Check the worker logs for more details on the error" to the top level exception's message in the REST API response (the worker logs will have the entire exception chain). Thoughts?

**Review by gharris1727 — APPROVED** [review]
LGTM, thanks for the fix!

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Unbounded waiting for the record to send doesn't seem very safe here. It looks like we were already trying to achieve synchronous writes by immediately performing a read-to-end after the call to `configLog::send` with a timeout of `READ_TO_END_TIMEOUT_MS`; could we use that same timeout while waiting for producer send futures to be completed (possibly after renaming to reflect the new purpose of the timeout)?

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Also, making this synchronously await the producer future's completion has the potential to slow things down for operations that involve writing multiple records to the config topic (such as [deleting a connector](https://github.com/apache/kafka/blob/b2cb546fba03bbdc4054c7d120b0b2654c7cf34e/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java#L493-L494)).

This is already the case with exactly-once support enabled since we perform each write in its own transaction, but while we're in the neighborhood, if there's an easy way to group together producer sends before awaiting the completion of any of their futures, it'd be nice. Not going to block on this, though.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: Do we still see the cause of the failure to send in the REST response with the current changes? It looks like we're only going to include the generic "Failed to... to/from Kafka" message, or the privileged write failure message.

I think it'd be nice to include more detail on the cause of the failure, but if that's too invasive or difficult to get quite right, it can be left as a follow-up item.

**Review by C0urante — COMMENTED** [review]
Thanks Yash! Left a few comments.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Thanks Chris, both great points. The `get` without timeout here was definitely a miss on my part. I've addressed both of your raised concerns in the latest patch (including batching multiple sends in a single transaction for the EOS enabled case).

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: Yeah, that's right, the API response will only include the generic top level message. 

> I think it'd be nice to include more detail on the cause of the failure

I strongly agree, and this was discussed in some more detail on the [other thread](https://github.com/apache/kafka/pull/12984#discussion_r1064077119).

> We wouldn't be making it more vague. The message would state that the write to the config topic failed which is the cause for failure. Since the exception mapper used by Connect's REST server only writes the [top level exception's message](https://github.com/apache/kafka/blob/d798ec779c25dba31fa5ee9384d159ed54c6e07b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/errors/ConnectExceptionMapper.java#L72) to the response (i.e. nested exceptions aren't surfaced via the REST API response), I think it makes sense to keep the top level exception's message generic and allow users to debug further via the worker logs (where the entire exception chain's stack trace will be visible).
...
The reasoning here is that since a Connect user may not even know that Connect uses a producer under the hood to write certain requests to the config topic for asynchronous processing, it would make more sense to have an informative Connect specific exception message rather than directly throwing the producer exception which may or may not contain enough details to be relevant to a Connect user.

> Another option for the above issue could be changing the exception mapper to concatenate all the exception messages from the exception chain.

> Yet another option for this could be to simply append a "Check the worker logs for more details on the error" to the top level exception's message in the REST API response (the worker logs will have the entire exception chain). Thoughts?

What do you think about modifying the exception mapper to be more informative (either in this PR or a separate one)?

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: Oof, that's a lot of context 😄

Thinking about it some more, I'm hesitant to make significant changes to the exception mapper without a KIP since it's a crucial part of our API and there may be automated tooling (like K8s operators) built around the current behavior, and adding full stack traces and/or caused-by chains could make things less readable and human-friendly, especially for new users.

Asking people to check the worker logs isn't a terrible solution, though it might be a bit tricky to make sure that that message reaches users regardless of whether EOS source support is enabled or disabled.

An alternative could be to handle this case specially by instantiating an exception whose message contains information on its cause. For example, the message could be `"Failed to write task configs to Kafka. Caused by org.apache.kafka.common.errors.AuthorizationException: Not authorized to access topics: connect-configs"`.

Regardless, this can and probably should be left as a follow-up since it's its own can of worms and doesn't have to block this PR.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: This gives us 30 seconds for each write instead of 30 seconds total, and it doesn't take into account reading to the end of the log after writes have finished. Considering this is all taking place on the herder's tick thread, we should probably care about the difference.

We might be able to use the [Timer class](https://github.com/apache/kafka/blob/eb7f490159c924ca0f21394d58366c257998f52e/clients/src/main/java/org/apache/kafka/common/utils/Timer.java) to simplify some of this logic.

**Review by C0urante — COMMENTED** [review]
Thanks Yash, just a couple more thoughts.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: > Considering this is all taking place on the herder's tick thread, we should probably care about the difference.

Makes sense.

> We might be able to use the [Timer class](https://github.com/apache/kafka/blob/eb7f490159c924ca0f21394d58366c257998f52e/clients/src/main/java/org/apache/kafka/common/utils/Timer.java) to simplify some of this logic.

Thanks, that's a great suggestion and it looks like a perfect fit for the use case here. One thing I'd like to call out is that the existing `putTaskConfigs` implementation had multiple (3) reads to the end of the log with each having a 30 second timeout (thus potentially blocking for up to 90 seconds in total). With the latest pushed changes in this PR, the overall timeout across all reads and writes done by `putTaskConfigs` will now be 30 seconds. While I do believe that this is the right thing to do, I just wanted to call it out explicitly to make sure that we're on the same page!

File: clients/src/main/java/org/apache/kafka/common/utils/Timer.java
Comment (C0urante) [inline clients/src/main/java/org/apache/kafka/common/utils/Timer.java:164]: In the future, would you mind saving nonessential improvements like this for a dedicated PR? They tend to add noise to the diff and makes things harder to review.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:796]: We should update the timer before waiting on the first future, since in some rare cases `configLog::send` may actually block for a bit:

```suggestion
            timer.update();
            for (Future<RecordMetadata> future : producerFutures) {
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:786]: Thank you for taking care to call out that the timer is updated before returning; it's an important detail.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: I'm not sure this belongs in a `finally` block, especially since that introduces inconsistencies in how this method updates the timer depending on whether EOS source support is enabled. IMO it's fine to add inside the `try` block after the call to `fencableProducer::commitTransaction`

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:716]: (Resolving this convo as it's taking up way too much real estate in the GitHub UI)

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java]: Yeah, that's fine. The less time we spend blocking the tick thread, the better.

**Review by C0urante — COMMENTED** [review]
Looking pretty good, almost there!

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (yashmayya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:796]: Done, but just curious about when the producer send might block? The doc says that it should just essentially add the record to the record buffer and return?

File: clients/src/main/java/org/apache/kafka/common/utils/Timer.java
Comment (yashmayya) [reply clients/src/main/java/org/apache/kafka/common/utils/Timer.java:164]: Makes sense, but wouldn't it also be slightly strange to create a dedicated PR for such a trivial change? 😄

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:796]: There's a lot of history here, but https://issues.apache.org/jira/browse/KAFKA-3539 is a decent starting point to learn more about when `KafkaProducer::send` may block.

**Review by C0urante — APPROVED** [review]
LGTM, thanks Yash!

We can't backport this to 3.4 right now, but given the importance of this fix, I'd like to once we're beyond code freeze. Filed https://issues.apache.org/jira/browse/KAFKA-14674 to track.

**Comment by yashmayya** [comment]
Thanks Greg and Chris for the great reviews on this one!

**Comment by C0urante** [comment]
Finished backporting to 3.3 and 3.4, now that both active releases on those branches have concluded.

**Pull request merged**
